File: module/lua_juce_core/text/String.cpp
namespace lua_juce {
auto juce_String(sol::table& state) -> void
{
    auto str = state.new_usertype<juce::String>("String", sol::constructors<juce::String(), juce::String(char const*)>());

    str["isEmpty"]     = LUA_JUCE_C_CALL(&juce::String::isEmpty);
    str["isNotEmpty"]  = LUA_JUCE_C_CALL(&juce::String::isNotEmpty);
    str["clear"]       = LUA_JUCE_C_CALL(&juce::String::clear);
    str["hashCode"]    = LUA_JUCE_C_CALL(&juce::String::hashCode);
    str["hashCode64"]  = LUA_JUCE_C_CALL(&juce::String::hashCode64);
    str["hash"]        = LUA_JUCE_C_CALL(&juce::String::hash);
    str["length"]      = LUA_JUCE_C_CALL(&juce::String::length);
    str["trim"]        = LUA_JUCE_C_CALL(&juce::String::trim);
    str["toStdString"] = LUA_JUCE_C_CALL(&juce::String::toStdString);
}
} // namespace lua_juce

File: tests/core/String.lua
local empty = juce.String.new()
assert(empty:isEmpty())
assert(not empty:isNotEmpty())
assert(empty:length() == 0)

local foo = juce.String.new("foo")
assert(not foo:isEmpty())
assert(foo:isNotEmpty())
assert(foo:length() == 3)

local long = juce.String.new("this variable tests a string that is long!")
assert(not long:isEmpty())
assert(long:isNotEmpty())
assert(long:length() == 42)
assert(long:hash() ~= 0)
assert(long:hashCode() ~= 0)
assert(long:hashCode64() ~= 0)

long:clear()
assert(long:isEmpty())
assert(not long:isNotEmpty())
assert(long:length() == 0)
